src/op_mode/vrrp.py

```diff
@@ -13,47 +13,308 @@
+   Gratuitous ARP delay: {{rec.garp_delay}}
+   Gratuitous ARP repeat: {{rec.garp_rep}}
+   Gratuitous ARP refresh: {{rec.garp_refresh}}
+   Gratuitous ARP refresh repeat: {{rec.garp_refresh_rep}}
+   Gratuitous ARP lower priority delay: {{rec.garp_lower_prio_delay}}
+   Gratuitous ARP lower priority repeat: {{rec.garp_lower_prio_rep}}
+   Send advert after receive lower priority advert: {{rec.lower_prio_no_advert}}
+   Send advert after receive higher priority advert: {{rec.higher_prio_send_advert}}
+   Virtual Router ID: {{rec.vrid}}
+   Priority: {{rec.base_priority}}
+   Effective priority: {{rec.effective_priority}}
+   Advert interval: {{rec.adver_int}} sec
+   Accept: {{rec.accept}}
+   Preempt: {{rec.nopreempt}}
+   {% if rec.preempt_delay -%}
+   Preempt delay: {{rec.preempt_delay}}
+   {% endif -%}
+   Promote secondaries: {{rec.promote_secondaries}}
+   Authentication type: {{rec.auth_type}}
+   {% if rec.vips %}
+   Virtual IP ({{ rec.vips | length }}):
+       {% for ip in rec.vips -%}
+         {{ip}}
+       {% endfor -%}
+   {% endif -%}
+   {% if rec.evips %}
+   Virtual IP Excluded:
+       {% for ip in rec.evips -%}
+         {{ip}}
+       {% endfor -%}
+   {% endif -%}
+   {% if rec.vroutes %}
+   Virtual Routes:
+       {% for route in rec.vroutes -%}
+         {{route}}
+       {% endfor -%}
+   {% endif -%}
+   {% if rec.vrules %}
+   Virtual Rules:
+       {% for rule in rec.vrules -%}
+         {{rule}}
+       {% endfor -%}
+   {% endif -%}
+   {% if rec.track_ifp %}
+   Tracked interfaces:
+       {% for ifp in rec.track_ifp -%}
+         {{ifp}}
+       {% endfor -%}
+   {% endif -%}
+   {% if rec.track_script %}
+   Tracked scripts:
+       {% for script in rec.track_script -%}
+         {{script}}
+       {% endfor -%}
+   {% endif %}
+   Using smtp notification: {{rec.smtp_alert}}
+   Notify deleted: {{rec.notify_deleted}}
+{% endfor %}
+""")
+
+# https://github.com/acassen/keepalived/blob/59c39afe7410f927c9894a1bafb87e398c6f02be/keepalived/include/vrrp.h#L126
+VRRP_AUTH_NONE = 0
+VRRP_AUTH_PASS = 1
+VRRP_AUTH_AH = 2
+
+# https://github.com/acassen/keepalived/blob/59c39afe7410f927c9894a1bafb87e398c6f02be/keepalived/include/vrrp.h#L417
+VRRP_STATE_INIT = 0
+VRRP_STATE_BACK = 1
+VRRP_STATE_MAST = 2
+VRRP_STATE_FAULT = 3
+
+VRRP_AUTH_TO_NAME = {
+    VRRP_AUTH_NONE: "NONE",
+    VRRP_AUTH_PASS: "SIMPLE_PASSWORD",
+    VRRP_AUTH_AH: "IPSEC_AH",
+}
+
+VRRP_STATE_TO_NAME = {
+    VRRP_STATE_INIT: "INIT",
+    VRRP_STATE_BACK: "BACKUP",
+    VRRP_STATE_MAST: "MASTER",
+    VRRP_STATE_FAULT: "FAULT",
+}
+
+
+def _get_raw_data(group_name: str = None) -> list:
+    """
+    Retrieve raw JSON data for all VRRP groups.
+
+    Args:
+        group_name (str, optional): If provided, filters the data to only
+            include the specified vrrp group.
+
+    Returns:
+        list: A list of raw JSON data for VRRP groups, filtered by group_name
+            if specified.
+    """
+    output = VRRP.collect('json')
```

Comment:
This function may raise `VRRPNoData`, so we need to catch it and re-raise it as `vyos.opmode.DataUnavailable`.

Reply:
fixed